[Hosting] Register ImageSource services as KeyedServices

mdh1418 opened the pull request:
### Description of Change

This PR aims to overhaul the [`ImageSourceServiceProvider`-based ImageSource Service registration setup](https://github.com/dotnet/maui/blob/61ca4146551affb9d56e2f165f4aaaac4c711a3d/src/Core/src/Hosting/ImageSources/ImageSourcesMauiAppBuilderExtensions.cs#L31-L32) in favor of leveraging [Keyed Services](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-8#keyed-di-services) introduced in .NET 8 to achieve a more trimmer friendly solution.

`ImageSourceServiceProvider` was initially introduced to be able to retrieve various derivations of ImageSourceService of which each would interact with its own derivation of an ImageSource (i.e. `FileImageSourceService : ImageSourceService, IImageSourceService<IFileImageSource>`). Thus, for the ImageSourceServiceProvider to grab these services, it would need to resolve an ImageSource to the corresponding `IImagesourceService<>` generic that the service had been registered with via trimming unfriendly `System.Type.MakeGenericType` https://github.com/dotnet/maui/blob/61ca4146551affb9d56e2f165f4aaaac4c711a3d/src/Core/src/Hosting/ImageSources/ImageSourceServiceProvider.cs#L25-L59.

Instead, leveraging Keyed Services to register and retrieve each `ImageSourceService` with a service key such as the underlying ImageSource type `IFileImageSource`, renders the logic to extrapolate an `IImageSourceService<>` generic from an ImageSource type unnecessary (one could grab the ImageSourceService via `GetRequiredKeyedService<IImageSourceService>(typeof(IFileImageSource))`). As a result, the `ImageSourceService`s can be registered directly through the MauiApp's ServiceCollection, making the ImageSourceServiceProvider class unnecessary.

This PR overhauls the `ImageSourceServiceProvider`-based ImageSource Service registration setup by:
- Registering the various ImageSource Services directly through the MauiApp's ServiceCollection
- Replacing ImageSourceServiceProvider API calls with the corresponding `GetRequiredKeyedService`/`GetKeyedService` calls
  - Ensuring providers used to retrieve ImageSourceServices can be casted to IKeyedServiceProvider
    - Modifying MauiContext's underlying WrappedServiceProvider implements IKeyedServiceProvider
    - Modifying Public API signatures to use IKeyedServiceProvider instead of IServiceProvider
- Removing ImageSourceServiceProvider all together
- [WIP] Modifying tests to reflect the change in ImageSourceService registration

### Issues Fixed

Fixes `ImageSourceServiceProvider` trimming warnings mentioned in https://github.com/dotnet/maui/issues/19397. Once this PR is merged, the number of warnings tracked by `PublishNativeAOTCheckWarnings` introduced in https://github.com/dotnet/maui/pull/19194 should decrease.

mdh1418 committed:
[MauiContext] Extend WrappedServiceProvider with IKeyedServiceProvider interface

mdh1418 committed:
Migrate UpdateFlyoutIconAsync to Keyed services

mdh1418 committed:
Migrate UpdateBackgroundImageSourceAsync to Keyed Services

mdh1418 committed:
Migrate UpdateThumbImageSourceAsync to Keyed Services

mdh1418 committed:
Migrate UpdateSourceAsync to KeyedServices

mdh1418 committed:
Migrate SetMenuItemIcon to Keyed Services

mdh1418 committed:
Migrate GetPlatformImageAsync to Keyed Services

mdh1418 committed:
Remove remaining references to IImageSourceServiceProvider

mdh1418 committed:
Remove ImageSourceServiceProvider

simonrozsival commented:
/azp run

rmarinho commented:
Can we get a little more context about this on the PR description?

mdh1418 commented:
@rmarinho Yes, sorry it is still a work in progress, and I haven't had time to fill out the description yet.

jonathanpeppers commented:
Can you target the `net9.0` branch instead? `main` is for .NET 8 servicing right now.

mdh1418 committed:
[ImageSourceService] Alias ImageSource handlers to ImageSourceServices

mdh1418 committed:
[ImageSourceService] Register ImageSource Services as Keyed services

With keyed services introduced in net8.0, register the default IImageSourceServices
as keyed services directly through the MauiApp's ServiceProvider's ServiceCollection
in preparation to obviate ImageSourceServiceProvider logic.

mdh1418 committed:
[ElementHandler] Add GetKeyedServiceProvider extension method

mdh1418 committed:
[ImageSourceService] Deprecate ConfigureImageSources with configureDelegate

mdh1418 committed:
[DeviceTests] Transition ConfigureTestBuilder to Keyed Service registration

mdh1418 committed:
[DeviceTests] Extend ContextStub to implement IKeyedServiceProvider

Tests leverage ContextStub instead of MauiContext. With ImageSourceServices
registered as keyed services, ContextStub needs to implement IKeyedServiceProvider
in order for tests to properly grab keyed image sources.

mdh1418 committed:
[DeviceTests] Alias FileImageSourceStub

The previous ImageSourceService registration setup leveraged reflection
to find any ImageSourceServices registered under a type's interface ancestry.
With the keyed services implementation, FileImageSourceStub needs to be aliased
to the registered FileImageSourceService for tests involving updates to
FileImageSourceStub to work properly.

mdh1418 committed:
[DeviceTests] Migrate ImageHandlerTests to keyed services

mdh1418 committed:
[DeviceTests] Migrate FontImageSourceServiceTests to Keyed Services

mdh1418 committed:
[DeviceTests] Migrate ImageSourceServiceTests to Keyed Services

mdh1418 committed:
[UnitTests] Migrate HostBuilderImageSourceTests to keyed services

mdh1418 committed:
[UnitTests] Migrate ImageSourceServiceTests to Keyed Services

mdh1418 committed:
[IntegrationTests] Remove ImageSourceServiceProvider warnings

mdh1418 committed:
Fixup UpdateBackgroundImageSourceAsync

mdh1418 commented:
/azp run

mdh1418 commented:
/azp run

jonathanpeppers commented on src/Compatibility/ControlGallery/src/Core/Startup.cs:
```diff
@@ -34,10 +34,7 @@ public static MauiAppBuilder CreateMauiAppBuilder()
-				{
-					sources.AddCompatibilityServices(AppDomain.CurrentDomain.GetAssemblies());
-				})
+				.AddCompatibilityServices(AppDomain.CurrentDomain.GetAssemblies())
```
This used to be lazy, where it wouldn't call `AppDomain.CurrentDomain.GetAssemblies()` until you called `Build()` on the "MauiAppBuilder".

Does this fix something, or should we change it back?

mdh1418 replied:
I removed `ConfigureImageSources(this, configureDelegate)` all together so ImageSourceServiceProvider could be removed completely, so I wanted to keep the AddCompatibilityServices logic.

mdh1418 replied:
Is there another way to keep the logic as lazy if we are removing ImageSourceServiceProvider? the lazy logic worked previously because the ImageSourceServiceProvider would only be initialized once there was a call to `GetRequiredService<IImageSourceServiceProvider>()` right? This PR currently has the keyed services being registered directly to the MauiApp ServiceProvider, so no intermediate ImageSourceServiceProvider.

jonathanpeppers replied:
For now, I think leave it as you have it here. We don't think there are many that would use this API.

jonathanpeppers commented on src/Controls/src/Core/Platform/Android/BottomNavigationViewUtils.cs:
```diff
@@ -132,9 +132,10 @@ internal static async Task SetMenuItemIcon(IMenuItem menuItem, ImageSource sourc
-			var services = context.Services;
-			var provider = services.GetRequiredService<IImageSourceServiceProvider>();
-			var imageSourceService = provider.GetRequiredImageSourceService(source);
+			if (context.Services is not IKeyedServiceProvider provider)
+				throw new InvalidOperationException($"Context services {nameof(context.Services)} does not implement {nameof(IKeyedServiceProvider)}");
```
@PureWeen can/should we change `context.Services` to an `IKeyedServiceProvider`?  (it extends `IServiceProvider`) Then we wouldn't need to cast/throw here?

https://github.com/dotnet/runtime/blob/0823c5c2814ffad5718f06715899ef91aa84aae2/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/IKeyedServiceProvider.cs#L12

The cast/throw occurs a couple places in this PR.

mdh1418 committed:
Fixup HostBuilderImageSourceTests

mdh1418 committed:
Fixup ImageSourceServiceTests

mdh1418 committed:
[DeviceTests] Transition Controls ConfigureTestBuilder to Keyed Service registration

mdh1418 committed:
[DeviceTests] Alias FileImageSource for Windows tests

mdh1418 committed:
Revert "[ImageSourceService] Deprecate ConfigureImageSources with configureDelegate"

This reverts commit d3e0305fd81827b74d98dd9c0c2ac724ea065666.

mdh1418 committed:
Revert "Remove ImageSourceServiceProvider"

This reverts commit cb0c1158a93c00ab0c9d082773817359c9729e22.

mdh1418 committed:
Revert "Remove remaining references to IImageSourceServiceProvider"

This reverts commit f840ac387479df694bd2da36ced8cac22db7cafd.

mdh1418 committed:
[ImageSourceServices] Transition ConfigureImageSources with delegate to keyed services

mdh1418 committed:
Transition AddServices to keyed services registration

mdh1418 committed:
Allow ConfigureImageSources calls to opt into keyed service transition
